lib/ansible/utils/collection_loader/_collection_finder.py

```diff
@@ -377,6 +456,29 @@ def _module_file_from_path(leaf_name, path):
 
         return module_path, has_code, package_path
 
+    def exec_module(self, module):
+        # execute the module's code in its namespace
+        code_obj = self.get_code(self._fullname)
+        if code_obj is not None:  # things like NS packages that can't have code on disk will return None
+            exec(code_obj, module.__dict__)
+
+    def create_module(self, spec):
+        # short-circuit redirect; we've already imported the redirected module, so just alias it and return it
+        if self._redirect_module:
+            sys.modules[self._fullname] = self._redirect_module
+            return self._redirect_module
+
+        module = ModuleType(spec.name)
+        module.__loader__ = spec.loader
+        module.__package__ = spec.parent
+        if spec.has_location:
+            module.__file__ = spec.origin
+        if spec.submodule_search_locations is not None:
+            module.__path__ = spec.submodule_search_locations
+
+        sys.modules[spec.name] = module
```

Comment:
I don't believe `create_module` should be adding to `sys.modules`.  The `importlib` documentation isn't clear here, but reading pep489 leads me to believe this should only happen during the `Module Execution Phase`.

I suppose it doesn't really matter in our context, since I doubt it's being used independently, but for correctness I believe it should be moved to `exec_module`.

`create_module` should only create the module shell, and the code isn't actually loaded until `exec_module`.  Which means that `create_module` shouldn't be loading any resources from the collection yet.

This creates a problem with `_AnsibleCollectionPkgLoader.create_module` though, since I think it is doing too much at that point.  I think `_AnsibleCollectionPkgLoader.create_module` probably needs removed and it's logic moved into an `exec_module`, or at minimum `AnsibleCollectionConfig.on_collection_load.fire` moved out of `create_module`, otherwise we get a recursion error.

I'd like to see what @nitzmahone thinks here.

Reply:
I can't think of any reason off the top of my head that we'd need to customize the module object creation, so just having create_module return `None` *should* be fine. 

IIRC `exec_module` is literally just about populating the module from the code; the 3.x import machinery takes care of the transactional insertion/removal from `sys.modules` on fresh imports and reloads, so we should probably never call that in the 3.x loader code at all. It's been awhile, but I think it basically just calls create_module (and does the stock creation if we didn't return something), provisionally inserts the module, calls exec_module under an exception handler, and remove the module on an error if it inserted it.

Reply:
I removed all of the extra sys.modules stuff, except from the redirect loader (which is maybe wrong?). There's a comment in importlib that seems like it supports create_module/exec_module modifying sys.modules at least, but maybe I'm making it overly complicated.
```
        _bootstrap._exec(spec, module)
        # The module may have replaced itself in sys.modules!
        return sys.modules[name]
```